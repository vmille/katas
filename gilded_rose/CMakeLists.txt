add_library(gilded_rose_lib)
target_sources(gilded_rose_lib PRIVATE src/gilded_rose.cxx)
target_include_directories(gilded_rose_lib PUBLIC include)

add_executable(gilded_rose_test)
target_sources(gilded_rose_test PRIVATE test/test_main.cxx test/test_cases.cxx)
target_link_libraries(gilded_rose_test PRIVATE gilded_rose_lib Catch2::Catch2 Catch2::Catch2WithMain ApprovalTests::ApprovalTests)