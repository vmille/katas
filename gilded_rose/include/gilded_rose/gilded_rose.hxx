//
// Created by Vivien Mille on 25/08/2024.
//

#ifndef KATAS_GILDED_ROSE_HXX
#define KATAS_GILDED_ROSE_HXX

#include <string>
#include <vector>

namespace vm {
  struct item {
    std::string name;
    int sell_in;
    int quality;

    item (std::string _name, int _sell_in, int _quality) : name(std::move(_name)), sell_in(_sell_in),
                                                           quality(_quality) {}
  };

  static constexpr auto aged_brie = "Aged Brie";
  static constexpr auto backstage_passes = "Backstage passes to a TAFKAL80ETC concert";
  static constexpr auto sulfuras = "Sulfuras, Hand of Ragnaros";
  static constexpr auto conjured = "Conjured";

  struct gilded_rose {
    std::vector<item>& items;

    gilded_rose (std::vector<item>& items);

    void update_quality ();
  };
}
#endif //KATAS_GILDED_ROSE_HXX
