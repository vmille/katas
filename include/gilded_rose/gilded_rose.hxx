//
// Created by Vivien Mille on 25/08/2024.
//

#ifndef KATAS_GILDED_ROSE_HXX
#define KATAS_GILDED_ROSE_HXX

#include <string>
#include <vector>

namespace vm {
    struct item {
        std::string name;
        int sellIn;
        int quality;
        item(std::string _name, int _sellIn, int _quality) : name(std::move(_name)), sellIn(_sellIn), quality(_quality) {}
    };

    struct gilded_rose {
        std::vector<item>& items;

        gilded_rose(std::vector<item>& items);

        void update_quality();
    };
}
#endif //KATAS_GILDED_ROSE_HXX
