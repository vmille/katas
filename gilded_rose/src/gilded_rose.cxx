//
// Created by Vivien Mille on 25/08/2024.
//

#include "gilded_rose/gilded_rose.hxx"

namespace vm {

  gilded_rose::gilded_rose (std::vector<item>& items) : items(items) {}

  void gilded_rose::update_quality () {
    for (int i = 0; i < items.size(); i++) {
      if (items[i].name != aged_brie && items[i].name != backstage_passes) {
        if (items[i].quality > 0 && items[i].name != sulfuras) {
          items[i].quality = items[i].quality - 1;
        }
      }
      else {
        if (items[i].quality < 50) {
          items[i].quality = items[i].quality + 1;
          if (items[i].name == backstage_passes) {
            if (items[i].sell_in < 11 && items[i].quality < 50) {
              items[i].quality = items[i].quality + 1;
            }

            if (items[i].sell_in < 6 && items[i].quality < 50) {
              items[i].quality = items[i].quality + 1;
            }
          }
        }
      }

      if (items[i].name != sulfuras) {
        items[i].sell_in = items[i].sell_in - 1;
      }

      if (items[i].sell_in < 0) {
        if (items[i].name != aged_brie) {
          if (items[i].name != backstage_passes) {
            if (items[i].quality > 0 && items[i].name != sulfuras) {
              items[i].quality = items[i].quality - 1;
            }
          }
          else {
            items[i].quality = items[i].quality - items[i].quality;
          }
        }
        else if (items[i].quality < 50) {
          items[i].quality = items[i].quality + 1;
        }
      }
    }
  }
} // vm