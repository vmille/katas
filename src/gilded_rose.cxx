//
// Created by Vivien Mille on 25/08/2024.
//

#include "gilded_rose/gilded_rose.hxx"

namespace vm {

    gilded_rose::gilded_rose(std::vector<item>& items) : items(items)
    {}

    void gilded_rose::update_quality()
    {
        for (int i = 0; i < items.size(); i++)
        {
            if (items[i].name != "Aged Brie" && items[i].name != "Backstage passes to a TAFKAL80ETC concert")
            {
                if (items[i].quality > 0)
                {
                    if (items[i].name != "Sulfuras, Hand of Ragnaros")
                    {
                        items[i].quality = items[i].quality - 1;
                    }
                }
            }
            else
            {
                if (items[i].quality < 50)
                {
                    items[i].quality = items[i].quality + 1;

                    if (items[i].name == "Backstage passes to a TAFKAL80ETC concert")
                    {
                        if (items[i].sell_in < 11)
                        {
                            if (items[i].quality < 50)
                            {
                                items[i].quality = items[i].quality + 1;
                            }
                        }

                        if (items[i].sell_in < 6)
                        {
                            if (items[i].quality < 50)
                            {
                                items[i].quality = items[i].quality + 1;
                            }
                        }
                    }
                }
            }

            if (items[i].name != "Sulfuras, Hand of Ragnaros")
            {
                items[i].sell_in = items[i].sell_in - 1;
            }

            if (items[i].sell_in < 0)
            {
                if (items[i].name != "Aged Brie")
                {
                    if (items[i].name != "Backstage passes to a TAFKAL80ETC concert")
                    {
                        if (items[i].quality > 0)
                        {
                            if (items[i].name != "Sulfuras, Hand of Ragnaros")
                            {
                                items[i].quality = items[i].quality - 1;
                            }
                        }
                    }
                    else
                    {
                        items[i].quality = items[i].quality - items[i].quality;
                    }
                }
                else
                {
                    if (items[i].quality < 50)
                    {
                        items[i].quality = items[i].quality + 1;
                    }
                }
            }
        }
    }
} // vm