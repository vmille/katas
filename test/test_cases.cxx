//
// Created by Vivien Mille on 25/08/2024.
//
#include <catch2/catch_test_macros.hpp>
#include <ApprovalTests/ApprovalTests.hpp>

#include "gilded_rose/gilded_rose.hxx"

#include <iostream>

namespace vm {
  std::ostream& operator<< (std::ostream& os, vm::item const& obj) {
    return os
        << "name: " << obj.name
        << ", sellIn: " << obj.sellIn
        << ", quality: " << obj.quality;
  }
}
// This is a normal unit test using Catch2
TEST_CASE("UpdateQuality") {
  std::vector<vm::item> items;
  items.push_back(vm::item("foo", 0, 0));
  vm::gilded_rose app(items);
  app.update_quality();
  REQUIRE("foo" == app.items[0].name);
}

// This is an Approval test using https://github.com/approvals/approvaltests.cpp
TEST_CASE("UpdateQualityApprovalTest") {
  std::vector<vm::item> items;
  items.push_back(vm::item("foo", 0, 0));
  vm::gilded_rose app(items);
  app.update_quality();
  auto item = app.items[0];
  ApprovalTests::Approvals::verify(item);
}

TEST_CASE("GildedRoseApprovalTests", "VerifyCombinations")
{
  std::vector<std::string> names{"Foo"};
  std::vector<int> sellIns{1};
  std::vector<int> qualities{1};

  auto f = [] (std::string name, int sellIn, int quality) {
    std::vector<vm::item> items = {vm::item(std::move(name), sellIn, quality)};
    vm::gilded_rose app(items);
    app.update_quality();
    return items[0];
  };

  ApprovalTests::CombinationApprovals::verifyAllCombinations(
      f,
      names, sellIns, qualities);
}