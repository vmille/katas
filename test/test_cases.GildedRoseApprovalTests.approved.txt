(Foo, 1, 1) => name: Foo, sell_in: 0, quality: 0

